presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayConcatUtils.java

```diff
@@ -25,7 +25,15 @@ private ArrayConcatUtils() {}
 
     public static Block concat(Type elementType, Block leftBlock, Block rightBlock)
     {
-        BlockBuilder blockBuilder = elementType.createBlockBuilder(new BlockBuilderStatus(), leftBlock.getSizeInBytes() + rightBlock.getSizeInBytes());
+        int totalSize = leftBlock.getSizeInBytes() + rightBlock.getSizeInBytes();
+        int totalEntries = leftBlock.getPositionCount() + rightBlock.getPositionCount();
+        if (leftBlock.getPositionCount() == 0) {
+            return rightBlock;
+        }
+        if (rightBlock.getPositionCount() == 0) {
+            return leftBlock;
+        }
+        BlockBuilder blockBuilder = elementType.createBlockBuilder(new BlockBuilderStatus(), totalEntries, (int) Math.ceil(totalSize / (double) totalEntries));
```

Comment:
I find increasing need of a method that takes number of entries and total size. Should the current `createBlockBuilder` be replaced?